### PR TITLE
Initial attempt to add support for the Laplace family

### PR DESCRIPTION
**Pull request by markgoodhead**
This is to address https://github.com/bambinos/bambi/issues/523

Note this is a WIP as I attempt to learn more of the bambi internals.

I'm not very familiar with the codebase but I attempted to copy the format of the Normal family and modify it accordingly. One thing I couldn't quite see is how this interacts with pymc - there's no "pm.Normal" calls which I was expecting to modify to the Laplace equivalent. Please let me know if there's something I'm missing and I'll attempt to fix this up.

### CODE REVIEW

**Comment by tomicapretto** [comment]
@markgoodhead thanks for opening this PR!

Initially this looks good! The only issue I would address is the name of the scale parameter in the Laplace distribution. In some places you're using `b` while in others you're using `scale`. From [the docs](https://docs.pymc.io/en/latest/api/distributions/generated/pymc.Laplace.html) I see the correct name is `b`.

File: bambi/defaults/defaults.py [hunk 1]
```diff
@@ -117,6 +119,17 @@
         "link": "log",
         "family": NegativeBinomial,
     },
+    "laplace": {
+        "likelihood": {
+            "name": "Laplace",
+            "args": {
+                "scale": "HalfNormal"
```
Comment (tomicapretto) [inline bambi/defaults/defaults.py:126]: `"scale"` should be `"b"`

File: bambi/families/likelihood.py [hunk 2]
```diff
@@ -15,6 +15,7 @@
     "Multinomial": DistSettings(params=("p",), parent="p", args=None),
     "Normal": DistSettings(params=("mu", "sigma"), parent="mu", args=("sigma",)),
     "NegativeBinomial": DistSettings(params=("mu", "alpha"), parent="mu", args=("alpha",)),
+    "Laplace": DistSettings(params=("mu", "scale"), parent="mu", args=("scale",)),
```
Comment (tomicapretto) [inline bambi/families/likelihood.py:18]: `"scale"` should be `"b"`

File: bambi/families/univariate.py [hunk 3]
```diff
@@ -90,6 +90,17 @@ def posterior_predictive(self, model, posterior, linear_predictor):
         return np.random.negative_binomial(n, p)
 
 
+class Laplace(UnivariateFamily):
+    SUPPORTED_LINKS = ["identity", "log", "inverse"]
+
+    def posterior_predictive(self, model, posterior, linear_predictor):
+        "Sample from posterior predictive distribution"
+        mean = self.link.linkinv(linear_predictor)
+        scale = posterior[model.response.name + "_scale"].values
```
Comment (tomicapretto) [inline bambi/families/univariate.py:99]: Here you also have to rename `"_scale"` with `"_b"`

**Review by tomicapretto — CHANGES_REQUESTED** [review]

**Comment by tomicapretto** [comment]
@markgoodhead for the missing `pm.Normal` call, you're right. Bambi implements a very general solution that avoids direct calls to pm.Normal, pm.Beta, etc. etc... It uses the name of the Likelihood function within the Family to select the appropiate PyMC distribution. Once that distribution is selected, it's a `dist(...)` call. All that happens in the following chunk

https://github.com/bambinos/bambi/blob/762f30a995a383f7588110f4343b765ed939ec70/bambi/backend/terms.py#L267-L295

**Comment by markgoodhead** [comment]
Thanks very much @tomicapretto ! I'll have a look at those changes now. I also spotted there's some interaction with the centered/non-centered functionality going on here for the Normal link case:

https://github.com/bambinos/bambi/blob/762f30a995a383f7588110f4343b765ed939ec70/bambi/backend/terms.py#L134-L150

From my experiments before the Laplace link has similar trade-offs to the Normal in terms of when to use the centered/non-centered parameterisation so perhaps this would also need modifying to switch between Normal and Laplace to implement both the centered/non-centered versions of both?

**Comment by markgoodhead** [comment]
> @markgoodhead for the missing `pm.Normal` call, you're right. Bambi implements a very general solution that avoids direct calls to pm.Normal, pm.Beta, etc. etc... It uses the name of the Likelihood function within the Family to select the appropiate PyMC distribution. Once that distribution is selected, it's a `dist(...)` call. All that happens in the following chunk
> 
> https://github.com/bambinos/bambi/blob/762f30a995a383f7588110f4343b765ed939ec70/bambi/backend/terms.py#L267-L295

On this point - it looks like Laplace won't need any special casing so I wouldn't need to do any changes here? In which case I think the outstanding changes would be a) the example you requested (I assume you mean a notebook in the examples folder?) and b) whatever changes you suggest to ensure compatibility with the `self.noncentered` parameter?

**Comment by markgoodhead** [comment]
> > Thanks very much @tomicapretto ! I'll have a look at those changes now. I also spotted there's some interaction with the centered/non-centered functionality going on here for the Normal link case:
> > https://github.com/bambinos/bambi/blob/762f30a995a383f7588110f4343b765ed939ec70/bambi/backend/terms.py#L134-L150
> > 
> > From my experiments before the Laplace link has similar trade-offs to the Normal in terms of when to use the centered/non-centered parameterisation so perhaps this would also need modifying to switch between Normal and Laplace to implement both the centered/non-centered versions of both?
> 
> Good catch!
> 
> This is not an adjustment that happens only for the Normal family (i.e. using Normal likelihood function). This happens for all group-specific effects (i.e. the ones you get with `1 | group` and `variable | group` syntax) no matter the family. So when you add the laplace family (i.e. Laplace likelihood), we will have this centered/non-centered parametrization for its group-specific effects as well.

Ah yes of course - my mistake! 

In that case I think the only change outstanding is the example you requested. I took a look at the notebooks in the example folder and they each seem like quite detailed explanations of a particular type of statistical technique, whereas this is sort of more 'minor' (just another family) so I'm not sure if it's appropriate for me just to add a very small notebook here or if there's a better place the example should live?

**Comment by tomicapretto** [comment]
> > > Thanks very much @tomicapretto ! I'll have a look at those changes now. I also spotted there's some interaction with the centered/non-centered functionality going on here for the Normal link case:
> > > https://github.com/bambinos/bambi/blob/762f30a995a383f7588110f4343b765ed939ec70/bambi/backend/terms.py#L134-L150
> > > 
> > > From my experiments before the Laplace link has similar trade-offs to the Normal in terms of when to use the centered/non-centered parameterisation so perhaps this would also need modifying to switch between Normal and Laplace to implement both the centered/non-centered versions of both?
> > 
> > 
> > Good catch!
> > This is not an adjustment that happens only for the Normal family (i.e. using Normal likelihood function). This happens for all group-specific effects (i.e. the ones you get with `1 | group` and `variable | group` syntax) no matter the family. So when you add the laplace family (i.e. Laplace likelihood), we will have this centered/non-centered parametrization for its group-specific effects as well.
> 
> Ah yes of course - my mistake!
> 
> In that case I think the only change outstanding is the example you requested. I took a look at the notebooks in the example folder and they each seem like quite detailed explanations of a particular type of statistical technique, whereas this is sort of more 'minor' (just another family) so I'm not sure if it's appropriate for me just to add a very small notebook here or if there's a better place the example should live?

No need to add a notebook at this time.. Just something like the example in this PR https://github.com/bambinos/bambi/pull/490

**Comment by aloctavodia** [comment]
I agree that there is no need to add a notebook as part of this PR, opening an issue as a remainder that the example is missing should be enough. Additionally, instead of writing a new notebook, we can maybe extend this one https://bambinos.github.io/bambi/main/notebooks/t_regression.html

**Comment by markgoodhead** [comment]
OK I've got a similar example that (appears to me anyway) to work on this branch - is this what you were expecting?

```
import arviz as az
import bambi as bmb
import numpy as np
import pandas as pd

az.style.use("arviz-darkgrid")
rng = np.random.default_rng(0)

size = 1000
x = rng.normal(loc=10., scale=5., size=size)
data = pd.DataFrame(
    {
        "x": x,
        "y": rng.laplace(loc=x, size=size)
    }
)

bmb_model = bmb.Model("y ~ x", data, family="laplace")
idata = bmb_model.fit(random_seed=0)

bmb_model.predict(idata)
az.plot_trace(idata);
```
image

**Review by tomicapretto — APPROVED** [review]

**Comment by tomicapretto** [comment]
I forgot to add that it would be good to have a test for this new family. Something like the examples attached would be good for now

https://github.com/bambinos/bambi/blob/762f30a995a383f7588110f4343b765ed939ec70/bambi/tests/test_built_models.py#L546-L548

https://github.com/bambinos/bambi/blob/762f30a995a383f7588110f4343b765ed939ec70/bambi/tests/test_built_models.py#L461-L463

**Comment by markgoodhead** [comment]
> Thanks a lot for this contribution! I hope it was a nice experience to you too.
> 
> If I can ask one more thing it would be comments about how easy/hard it was to work through this PR and what we could do to help you. Thanks again!

Sure - I can't think of anything big that'd change to make the process easier. There's always friction learning a new codebase but this wasn't too bad - the code is well laid out and generally quite readable.

Only two quite minor things I can think of - you can add the guidelines (which mention black and pylint) to be displayed on a PR as you open it (or a link to them) as I didn't spot those until quite late. Also, I got caught when my version of black was too old and did some bad formatting, perhaps the versions of pylint and black used in github actions could be specified in the PR guidelines?

**Pull request merged**